File: packages/client/src/lib/audioQuality.ts
import type { AudioQuality, MusicSource, PlatformAuthStatus } from '@music-together/shared'

export interface AudioQualityOption {
  value: AudioQuality
  label: string
  platform?: MusicSource
  description?: string
}

const BASE_OPTIONS: AudioQualityOption[] = [
  { value: 128, label: '标准 128kbps' },
  { value: 192, label: '较高 192kbps' },
  { value: 320, label: '高品质 320kbps' },
]

const PLATFORM_OPTIONS: Record<MusicSource, AudioQualityOption[]> = {
  netease: [
    { value: 999, label: '无损 SQ', platform: 'netease' },
    { value: 'netease_hires', label: 'Hi-Res', platform: 'netease' },
    { value: 'netease_jyeffect', label: '高清臻音', platform: 'netease' },
    { value: 'netease_spatial', label: '沉浸环绕声', platform: 'netease' },
    { value: 'netease_master', label: '超清母带', platform: 'netease' },
    { value: 'netease_dolby', label: '杜比全景声', platform: 'netease', description: '浏览器兼容性有限' },
  ],
  tencent: [
    { value: 'tencent_flac', label: '无损', platform: 'tencent' },
    { value: 'tencent_master', label: '臻品母带', platform: 'tencent' },
  ],
  kugou: [
    { value: 'kugou_hires', label: '酷狗 Hi-Res', platform: 'kugou' },
    { value: 'kugou_master', label: '酷狗臻品母带', platform: 'kugou' },
  ],
  kugou_concept: [
    { value: 'kugou_hires', label: '酷狗概念版 Hi-Res', platform: 'kugou_concept' },
    { value: 'kugou_master', label: '酷狗概念版 臻品母带', platform: 'kugou_concept' },
  ],
  bilibili: [],
}

export function getAudioQualityOptions(_statuses: PlatformAuthStatus[]): AudioQualityOption[] {
  // Room quality is shared by all providers. Provider-specific formats are
  // selected by the server from `highest`; exposing them here made a Tencent
  // or Kugou room appear to offer unrelated Netease formats.
  return [
    ...BASE_OPTIONS,
    { value: 999, label: '无损 SQ', description: '需要 VIP 账号' },
    { value: 'highest', label: '尽量高' },
  ]
}

export function getAudioQualityLabel(quality: AudioQuality, statuses: PlatformAuthStatus[] = []): string {
  const allOptions = [...BASE_OPTIONS, ...Object.values(PLATFORM_OPTIONS).flat()]
  return (
    [...getAudioQualityOptions(statuses), ...allOptions].find((option) => option.value === quality)?.label ??
    String(quality)
  )
}

File: packages/android-client/app/src/main/java/io/github/yueby/musictogether/model/AudioQuality.kt
package io.github.yueby.musictogether.model

data class AudioQualityOption(
    val value: String,
    val label: String,
    val platform: String? = null,
    val description: String? = null,
)

private val baseAudioQualities = listOf(
    AudioQualityOption("128", "标准 128kbps"),
    AudioQualityOption("192", "较高 192kbps"),
    AudioQualityOption("320", "高品质 320kbps"),
)

private val legacyAudioQualities = listOf(
    AudioQualityOption("netease_hires", "Hi-Res", "netease"),
    AudioQualityOption("netease_jyeffect", "高清臻音", "netease"),
    AudioQualityOption("netease_spatial", "沉浸环绕声", "netease"),
    AudioQualityOption("netease_master", "超清母带", "netease"),
    AudioQualityOption("netease_dolby", "杜比全景声", "netease", "兼容性有限"),
    AudioQualityOption("tencent_flac", "QQ 无损", "tencent"),
    AudioQualityOption("tencent_master", "QQ 臻品母带", "tencent"),
    AudioQualityOption("kugou_hires", "酷狗 Hi-Res", "kugou"),
    AudioQualityOption("kugou_master", "酷狗蝰蛇母带 2.0", "kugou"),
    AudioQualityOption("bilibili_64", "B站 64K", "bilibili"),
    AudioQualityOption("bilibili_132", "B站 132K", "bilibili"),
    AudioQualityOption("bilibili_192", "B站 192K", "bilibili"),
    AudioQualityOption("bilibili_hires", "B站 Hi-Res", "bilibili"),
)

fun availableAudioQualities(): List<AudioQualityOption> = baseAudioQualities + listOf(
    AudioQualityOption("999", "无损 SQ", description = "需要 VIP 账号"),
    AudioQualityOption("highest", "尽量高"),
)

fun audioQualityLabel(value: String): String =
    (availableAudioQualities() + legacyAudioQualities)
        .firstOrNull { it.value == value }
        ?.label
        ?: value
